Improve public interface

samuelmtimbo opened the pull request:
'handler' option can now be a string or omitted;
'run()' now receives a root parameter;
Refactor code to conform with new paradigm;

vibegui commented on README.md:
```diff
@@ -139,23 +136,38 @@ login: {
+  new: {
+    requiredArgs: 'name',
+    description: 'Create a new workspace',
+    // will look './workspace/new' (from root) for handling function
```
`Will look at`

vibegui commented on README.md:
```diff
@@ -139,23 +136,38 @@ login: {
+        type: 'string',
+      },
+    ],
+    // will look './workspace/delete' (from root) for handling function
```
☝️

vibegui commented on package-lock.json:
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "findhelp",
```
😬  `yarn.lock` pls

vibegui commented on src/finder.js:
```diff
@@ -45,19 +50,11 @@ export function findOptions (node) {
 
 export function isCommand (node) {
-  return !isNil(node) && type(node.handler) === 'Function' && node
+  return !isNil(node) && (node.handler || node.options || node.alias || node.description) && node
```
Should also check for `requiredArgs`

vibegui commented on src/finder.js:
```diff
@@ -73,29 +70,34 @@ export function findNext (key, node) {
+  node.path = path
+  forEachObjIndexed((n, key) => {
+    if (isCommand(n) || isNamespace(n)) {
+      addPaths(n, `${path}/${key}`)
```
`find` is already recursive so there's no need to visit the whole tree from each node, just adding the path for the current node would suffice.

vibegui commented on src/finder.js:
```diff
@@ -109,6 +111,15 @@ export function find (node, args, raw, minimist) {
+  if (typeof command.handler === 'function') {
+    return command.handler.apply(this, args)
+  } else if (typeof command.handler === 'string') {
+    return require(path.join(root, command.handler)).default.apply(this, args)
```
Should use an util that checks for `req.default || req` for interop with `module.exports = function(){}`.

vibegui commented on src/finder.js:
```diff
@@ -109,6 +111,15 @@ export function find (node, args, raw, minimist) {
+  } else if (typeof command.handler === 'string') {
+    return require(path.join(root, command.handler)).default.apply(this, args)
+  } else if (typeof command.handler === 'undefined') {
+    console.log(command)
```
:eyes:

vibegui commented on src/finder.js:
```diff
@@ -109,6 +111,15 @@ export function find (node, args, raw, minimist) {
-  return command.handler.apply(this, args)
+export function run ({command, args}, root) {
+  if (typeof command.handler === 'function') {
+    return command.handler.apply(this, args)
```
`.apply(this, args)` is the same logic for every case. It seems you should just decide _how_ to find the command.

vibegui commented on src/finder.js:
```diff
@@ -109,6 +111,15 @@ export function find (node, args, raw, minimist) {
+    return require(path.join(root, command.path)).default.apply(this, args)
+  }
+
+  throw new Error('?') // TODO
```
- Throw an specific error for "handler is not a supported type, expected function, string or handler defined at `{{path}}`, found i.e. Number" (UnsupportedHandler?)
- We should surround the requires with `try/catch` and check for `Module not found`, rethrow specific error with info about where command should be found and wasn't. (HandlerNotFound?)

vibegui reviewed (changes requested):
Almost there :)

samuelmtimbo committed:
Improve public interface

'handler' option can now be a string or omitted;
'run()' now receives a root parameter;
Refactor code to conform with new paradigm;

vibegui reviewed (approved):
LGTM! We need to cut a new major for this one.
